tests/org.eclipse.n4js.ide.tests/src/org/eclipse/n4js/ide/tests/server/AbstractIdeTest.java

```diff
@@ -291,26 +352,101 @@ protected void changeNonOpenedFile(String moduleName, Function<String, String> m
+		changeOpenedFile(getFileURIFromModuleName(moduleName), newContent);
+	}
+
+	/**
+	 * Same as {@link #changeOpenedFile(FileURI, Function)}, but the desired new content is given as argument instead of
+	 * a modification function.
+	 */
+	protected void changeOpenedFile(FileURI fileURI, String newContent) {
+		changeOpenedFile(fileURI, oldContent -> newContent);
+	}
+
+	/** Same as {@link #changeFileOnDiskWithoutNotification(FileURI, Function)}, but accepts a module name. */
+	protected void changeOpenedFile(String moduleName, Function<String, String> modification) {
+		changeOpenedFile(getFileURIFromModuleName(moduleName), modification);
+	}
+
 	/**
-	 * Change an opened file on disk and notify the LSP server.
+	 * Similar to {@link #changeOpenedFile(FileURI, Pair...)}, but the file's entire content is replaced in a single
+	 * {@link TextDocumentContentChangeEvent} by the string returned from the given modification function.
+	 * <p>
+	 * NOTE: this method replaces the file's entire content in a single step; if a test wants to simulate manual edits
+	 * more closely, then method {@link #changeOpenedFile(FileURI, Pair...)} should be used instead!
+	 *
+	 * @param modification
+	 *            will be invoked with the file's old content as argument and is expected to return the desired new
+	 *            content.
+	 */
+	protected void changeOpenedFile(FileURI fileURI, Function<String, String> modification) {
+		changeOpenedFile(fileURI,
+				modification,
+				(oldContent, newContent) -> singletonList(new TextDocumentContentChangeEvent(newContent)));
+	}
+
+	/** Same as {@link #changeOpenedFile(FileURI, Pair...)}, but accepts a module name. */
+	@SafeVarargs
+	protected final void changeOpenedFile(String moduleName, Pair<String, String>... replacements) {
+		changeOpenedFile(getFileURIFromModuleName(moduleName), replacements);
+	}
+
+	/**
+	 * Change an opened file <u>in memory only</u> and notify the LSP server.
 	 * <p>
 	 * Use method {@link #changeNonOpenedFile(String, Pair...)} instead if the file was *not* previously opened with one
-	 * of the {@link #openFile(String, String) #openFile()} methods.
+	 * of the {@link #openFile(FileURI) #openFile()} methods.
 	 */
-	protected void changeOpenedFile(String moduleName,
-			@SuppressWarnings("unchecked") Pair<String, String>... replacements) {
+	@SafeVarargs
+	protected final void changeOpenedFile(FileURI fileURI, Pair<String, String>... replacements) {
+		changeOpenedFile(fileURI,
+				oldContent -> applyReplacements(oldContent, replacements),
+				(oldContent, newContent) -> replacementsToChangeEvents(oldContent, replacements));
+	}
 
-		FileURI fileURI = getFileURIFromModuleName(moduleName);
-		// 1) change on disk
-		Pair<String, String> oldToNewContent = changeFileOnDiskWithoutNotification(fileURI, replacements);
-		String oldContent = oldToNewContent.getKey();
+	private void changeOpenedFile(FileURI fileURI, Function<String, String> modification,
```

Comment:
I think it now makes sense to split/refactor this class to e.g. extract/delegate client related lsp simulation methods, because this class got quite large with this PR (and with one of the prior PRs regarding lsp client simulations).
(Similar to the refactoring before that was extracting some logic into `TestWorkspaceCreator`.) Also it might become reasonable to move some of the methods into the `AbstractIncrementalBuilderTest`.

Reply:
This is a good refactoring we should do at some point (we talked about it before). But I don't want to include it in this PR. Will create a ticket for upcoming clean ups / refactorings in the test infrastructure.